Content/NPCs/Moonstone/DreambeastProjectiles.cs

```diff
@@ -0,0 +1,258 @@
+
+			return false;
+		}
+
+		public void DrawAdditive(SpriteBatch spriteBatch)
+		{
+			Texture2D tex = ModContent.Request<Texture2D>(Texture + "_Bloom").Value;
+			var glowColor = new Color(78, 87, 191);
+
+			if (Main.LocalPlayer.GetModPlayer<LunacyPlayer>().Insane)
+				spriteBatch.Draw(tex, Projectile.Center - Main.screenPosition, tex.Frame(), glowColor * 0.8f, Projectile.rotation, tex.Size() / 2, 1.2f, 0, 0);
+		}
+	}
+
+	public class DreambeastProjHome : ModProjectile, IDrawPrimitive, IDrawAdditive
+	{
+		public ref float target => ref Projectile.ai[0];
+
+		private List<Vector2> cache;
+		private Trail trail;
+
+		public Player Target => Main.player[(int)target];
+
+		public Vector2 PathCenter
+		{
+			get => new(Projectile.ai[1], Projectile.ai[2]);
+			set
+			{
+				Projectile.ai[1] = value.X;
+				Projectile.ai[2] = value.Y;
+			}
+		}
+
+		public override string Texture => AssetDirectory.SquidBoss + "InkBlob";
+
+		public override void SetDefaults()
+		{
+			Projectile.hostile = true;
+			Projectile.width = 36;
+			Projectile.height = 20;
+			Projectile.penetrate = 1;
+			Projectile.timeLeft = 180;
+			Projectile.tileCollide = false;
+			Projectile.ignoreWater = true;
+			Projectile.damage = 66;
+		}
+
+		public override void SetStaticDefaults()
+		{
+			DisplayName.SetDefault("Dream Essence");
+		}
+
+		public override void OnSpawn(IEntitySource source)
+		{
+			PathCenter = Projectile.Center;
+		}
+
+		public override void AI()
+		{
+			float homingMult = Math.Min(1 - (Projectile.timeLeft - 120f) / 60f, 1);
+
+			Projectile.velocity = Vector2.Lerp(Projectile.velocity, PathCenter.DirectionTo(Target.position) * 15, 0.03f * homingMult);
+
+			Projectile.rotation += 0.05f;
+
+			PathCenter += Projectile.velocity;
+
+			float wave = (float)Math.Sin(Projectile.timeLeft / 5f) * 20f * homingMult;
+
+			Projectile.Center = PathCenter + Projectile.velocity.SafeNormalize(Vector2.Zero).RotatedBy(MathHelper.PiOver2) * wave;
+
+			var glowColor = new Color(78, 87, 191);
+			Lighting.AddLight(Projectile.Center, glowColor.ToVector3());
+
+			if (Main.rand.NextBool(10))
+			{
+				Dust.NewDust(Projectile.Center, 0, 0, ModContent.DustType<Dusts.GlowFastDecelerate>(), 0, 0, 35, new Color(150, 120, 255) * 0.5f, Main.rand.NextFloat(0.6f, 0.8f));
+				Dust.NewDust(Projectile.Center, 0, 0, ModContent.DustType<Dusts.MoonstoneShimmer>(), 0, 0, 35, new Color(150, 120, 255, 0) * 0.5f, Main.rand.NextFloat(0.4f, 0.5f));
+			}
+
+			if (!Main.dedServ)
+			{
+				ManageCaches();
+				ManageTrail();
+			}
+
+			if (Projectile.timeLeft < 30)
+				Projectile.Opacity -= 0.033f;
+		}
+
+		public override bool CanHitPlayer(Player target)
+		{
+			return Main.LocalPlayer.GetModPlayer<LunacyPlayer>().Insane;
+		}
+
+		public override void ModifyHitPlayer(Player target, ref Player.HurtModifiers modifiers)
+		{
+			target.GetModPlayer<LunacyPlayer>().ReturnSanity(4);
```

Comment:
remember that these onhits like this are clientside only. this might be correct as is but only if this "ReturnSanity" doesn't need to be synced

Reply:
you can use PlayerHitPacket in these modifyHitPlayers though you'll need to extend it to call modifyHitPlayer too